File: demos/cpp/sndfile_demo.cpp
// demos/sndfile_demo.cpp
//
// Audio processing demo: read a real song, analyze its spectrum,
// apply a low-pass filter, and write the result.
//
// Usage: ./demo_sndfile
// Output: /tmp/insight_demo_*.wav

#ifdef INSIGHT_USE_SNDFILE

#include "insight/core/array.h"
#include "insight/init.h"
#include "insight/io/sndfile.h"
#include "insight/ops/creation.h"
#include "insight/ops/elementwise.h"
#include "insight/ops/fft.h"
#include "insight/ops/operator.h"
#include "insight/ops/reduction.h"
#include "insight/ops/signal.h"
#include <cmath>
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>

using namespace ins;

static const std::string TMP = "/tmp/insight_demo_";

// Generate a multi-frequency test signal
static Array generate_test_signal(int64_t frames, float sample_rate) {
  std::vector<float> data(frames);
  for (int64_t i = 0; i < frames; ++i) {
    float t = static_cast<float>(i) / sample_rate;
    data[i] = 0.5f * std::sin(2.0f * M_PI * 440.0f * t) +
              0.3f * std::sin(2.0f * M_PI * 880.0f * t) +
              0.1f * std::sin(2.0f * M_PI * 3141.0f * t);
  }
  return to_array<float>(data);
}

int main() {
  ins::init({"cpu"});

  std::cout << "=== Insight Audio Processing Demo ===" << std::endl;
  std::cout << std::endl;

  // Try to read the real audio file first
  const std::string audio_path =
      AUDIO_DIR "/大梦何时了-叶炫清-210356449-2000.flac";

  Array signal;
  int sample_rate;
  int64_t frames;

  try {
    std::cout << "[0] Reading: " << audio_path << std::endl;
    auto [data, info] = audio::read(audio_path);
    sample_rate = info.samplerate;
    frames = info.frames;

    std::cout << "    Format: FLAC, " << info.channels << "ch, "
              << info.samplerate << "Hz, " << info.frames << " frames ("
              << static_cast<float>(info.frames) / info.samplerate << "s)"
              << std::endl;

    // Mix to mono by averaging channels
    if (info.channels == 2) {
      // data shape: [frames, 2]
      Array left = data.slice(1, 0, 1).squeeze(1);  // [frames]
      Array right = data.slice(1, 1, 2).squeeze(1); // [frames]
      signal = (left + right) * 0.5f;
    } else {
      signal = data.squeeze(1); // [frames]
    }
  } catch (const std::exception &e) {
    std::cout << "    Could not read FLAC file: " << e.what() << std::endl;
    std::cout << "    Generating synthetic signal instead..." << std::endl;
    sample_rate = 44100;
    frames = static_cast<int64_t>(sample_rate * 2);
    signal = generate_test_signal(frames, static_cast<float>(sample_rate));
  }

  // Only process first 5 seconds to keep it fast
  int64_t max_frames = static_cast<int64_t>(sample_rate) * 5;
  if (frames > max_frames) {
    std::cout << "    Truncating to first 5 seconds (" << max_frames
              << " frames)" << std::endl;
    signal = signal.slice(0, 0, max_frames);
    frames = max_frames;
  }

  std::cout << std::endl;

  // Step 1: Write mono signal
  std::cout << "[1] Writing mono signal..." << std::endl;
  Array signal_2d = signal.reshape({frames, 1});
  audio::write(TMP + "original.wav", signal_2d, sample_rate);
  std::cout << "    Written: " << TMP << "original.wav" << std::endl;

  // Step 2: Low-pass filter via FFT
  std::cout << "[2] Applying low-pass filter (cutoff 1000Hz)..." << std::endl;
  Array spectrum = fft::rfft(signal);

  int64_t freq_bins = spectrum.numel();
  int64_t cutoff_bin =
      static_cast<int64_t>(1000.0 / (sample_rate / 2.0) * freq_bins);
  int64_t taper_width = freq_bins / 100;

  // Build a complex mask: pass low frequencies, taper at cutoff
  std::vector<std::complex<float>> mask_data(freq_bins, {0.0f, 0.0f});
  for (int64_t i = 0; i < cutoff_bin - taper_width / 2; ++i) {
    mask_data[i] = {1.0f, 1.0f};
  }
  for (int64_t i = std::max(0L, cutoff_bin - taper_width / 2);
       i < cutoff_bin + taper_width / 2 && i < freq_bins; ++i) {
    float t = 1.0f - static_cast<float>(i - cutoff_bin + taper_width / 2) /
                         taper_width;
    float v = std::max(0.0f, std::min(1.0f, t));
    mask_data[i] = {v, v};
  }

  Array mask = to_array<std::complex<float>>(mask_data, spectrum.shape(),
                                             ins::CPUPlace());
  Array filtered_spectrum = mul(spectrum, mask);
  Array filtered = fft::irfft(filtered_spectrum, frames);

  // Reshape to [frames, 1] for write
  Array filtered_2d = filtered.reshape({frames, 1});
  audio::write(TMP + "filtered.wav", filtered_2d, sample_rate);
  std::cout << "    Written: " << TMP << "filtered.wav" << std::endl;

  // Step 3: Energy comparison
  std::cout << std::endl;
  std::cout << "[3] Signal statistics:" << std::endl;
  Array energy_before = sum(mul(signal, signal));
  Array energy_after = sum(mul(filtered, filtered));
  float eb = energy_before.item<float>();
  float ea = energy_after.item<float>();
  std::cout << "    Energy (original):  " << eb << std::endl;
  std::cout << "    Energy (filtered):  " << ea << std::endl;
  if (eb > 0) {
    std::cout << "    Retained: " << (ea / eb) * 100.0f << "%" << std::endl;
  }

  std::cout << std::endl;
  std::cout << "Output files:" << std::endl;
  std::cout << "  " << TMP << "original.wav  - Mono mixdown" << std::endl;
  std::cout << "  " << TMP << "filtered.wav  - Low-pass < 1000Hz" << std::endl;

  return 0;
}

#else

#include <iostream>

int main() {
  std::cerr << "This demo requires libsndfile (INSIGHT_USE_SNDFILE)."
            << std::endl;
  return 1;
}

#endif

File: demos/julia/sndfile_demo.jl
# demos/julia/sndfile_demo.jl
#
# Signal processing demo: generate a multi-frequency test signal,
# apply a low-pass filter via FFT, and save results.
# Mirrors the C++ sndfile_demo without libsndfile dependency.

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "bindings", "julia"))
using Insight

function separator(title)
    println("=" ^ 60)
    println("  $title")
    println("=" ^ 60)
end

Insight.init(["cpu"])

# Parameters
sample_rate = 44100
duration = 2.0
frames = Int(sample_rate * duration)

println("=== Insight Signal Processing Demo (Julia) ===")
println("    Sample rate: $sample_rate Hz")
println("    Duration:    $duration s ($frames frames)")
println()

# Step 1: Generate multi-frequency test signal
separator("[1] Generate test signal")
println("    Components: 440Hz (0.5), 880Hz (0.3), 3141Hz (0.1)")

signal_data = Float32[0.5 * sin(2π * 440 * i / sample_rate) +
                       0.3 * sin(2π * 880 * i / sample_rate) +
                       0.1 * sin(2π * 3141 * i / sample_rate)
                       for i in 0:frames-1]

signal = Insight.from_data(signal_data, [frames])
println("    Signal: [$(frames)] elements")
println()

# Step 2: Write signal to binary file
separator("[2] Save signal to binary file")
tmp = "/tmp/insight_demo_"

Insight.write_bin(tmp * "original.bin", signal)
println("    Written: $(tmp)original.bin")

read_back = Insight.read_bin(tmp * "original.bin")
println("    Roundtrip read: $(Insight.numel(read_back)) elements")
println()

# Step 3: FFT analysis
separator("[3] FFT analysis")

spectrum = Insight.fft(signal)
freq_bins = Insight.numel(spectrum)
println("    FFT bins: $freq_bins")
println("    Nyquist: $(div(sample_rate, 2)) Hz")
println("    Bin resolution: $(round(sample_rate / freq_bins, digits=2)) Hz")
println()

# Step 4: Low-pass filter via FFT (cutoff 1000Hz)
separator("[4] Low-pass filter (cutoff 1000Hz)")

cutoff_bin = Int(floor(1000 / (sample_rate / 2.0) * freq_bins))
taper_width = Int(floor(freq_bins / 100))
println("    Cutoff bin: $cutoff_bin (1000Hz)")
println("    Taper width: $taper_width bins")

# Zero out frequencies above cutoff
# Work with magnitude: keep bins below cutoff, taper at cutoff
filtered_spectrum = Insight.fft(signal)

# Create a frequency-domain mask by manipulating the spectrum
# For simplicity, use rfft approach: keep only low-frequency bins
r_spectrum = Insight.rfft(signal)
r_bins = Insight.numel(r_spectrum)

# Build mask for real FFT
mask_data = Float32[]
for i in 0:r_bins-1
    if i < cutoff_bin - div(taper_width, 2)
        push!(mask_data, 1.0f0)
    elseif i < cutoff_bin + div(taper_width, 2)
        t_val = 1.0f0 - (i - cutoff_bin + div(taper_width, 2)) / taper_width
        push!(mask_data, max(0.0f0, min(1.0f0, t_val)))
    else
        push!(mask_data, 0.0f0)
    end
end

mask = Insight.from_data(mask_data, [r_bins])

# Apply mask to magnitude of rfft result
r_real = Insight.real_part(r_spectrum)
r_imag = Insight.imag_part(r_spectrum)
filt_real = Insight.mul(r_real, mask)
filt_imag = Insight.mul(r_imag, mask)
filt_spectrum = Insight.to_complex(filt_real, filt_imag)

# Inverse FFT
filtered = Insight.irfft(filt_spectrum, frames)
println("    Filtered signal: [$(Insight.numel(filtered))] elements")

# Step 5: Energy comparison
println()
separator("[5] Signal statistics")

energy_orig = Insight.sum(Insight.mul(signal, signal))
energy_filt = Insight.sum(Insight.mul(filtered, filtered))

eo = Insight.item(energy_orig)
ef = Insight.item(energy_filt)

println("    Energy (original): $eo")
println("    Energy (filtered): $ef")
if eo > 0
    println("    Retained: $(round(ef / eo * 100, digits=1))%")
end

# Write filtered signal
Insight.write_bin(tmp * "filtered.bin", filtered)
println()
println("Output files:")
println("  $(tmp)original.bin  - Original multi-freq signal")
println("  $(tmp)filtered.bin  - Low-pass filtered (<1000Hz)")
println()
println("Done!")
